packages/ripple-keypairs/src/utils/getAlgorithmFromKey.ts

```diff
@@ -0,0 +1,121 @@
+ * | secp256k1 | Public (2)  |  0x03  |   33   | prefix + 32 x-bytes (y is odd)                        | ecdsa-secp256k1 |
+ * | secp256k1 | Public (3)  |  0x04  |   65   | prefix + 32 x-bytes + 32 y-bytes                      | ecdsa-secp256k1 |
+ * | secp256k1 | Private (1) |  None  |   32   | Uint256BE (0 < n < order)                             | ecdsa-secp256k1 |
+ * | secp256k1 | Private (2) |  0x00  |   33   | prefix + Uint256BE (0 < n < order)                    | ecdsa-secp256k1 |
+ *
+ * Note: The 0x00 prefix for secpk256k1 Private (2) essentially 0 pads the number
+ *       and the interpreted number is the same as 32 bytes.
+ */
+const KEY_TYPES: Record<CompositeKey, Algorithm> = {
+  [`private_${Prefix.NONE}_32`]: 'ecdsa-secp256k1',
+  [`private_${Prefix.SECP256K1_PRIVATE}_33`]: 'ecdsa-secp256k1',
+  [`private_${Prefix.ED25519}_33`]: 'ed25519',
+  [`public_${Prefix.ED25519}_33`]: 'ed25519',
+  [`public_${Prefix.SECP256K1_PUB_X}_33`]: 'ecdsa-secp256k1',
+  [`public_${Prefix.SECP256K1_PUB_X_ODD_Y}_33`]: 'ecdsa-secp256k1',
+  [`public_${Prefix.SECP256K1_PUB_XY}_65`]: 'ecdsa-secp256k1',
+}
+
+function getKeyInfo(key: HexString) {
+  return {
+    prefix: key.length < 2 ? Prefix.NONE : parseInt(key.slice(0, 2), 16),
+    len: key.length / 2,
+  }
+}
+
+function prefixRepr(prefix: Prefix): string {
+  return prefix === Prefix.NONE
+    ? 'None'
+    : `0x${prefix.toString(16).padStart(2, '0')}`
+}
+
+function getValidFormatsTable(type: KeyType) {
+  // No need overkill with renderTable method
+  const padding = 2
+  const colWidth = {
+    algorithm: 'ecdsa-secp256k1'.length + padding,
+    prefix: '0x00'.length + padding,
+  }
+
+  return Object.entries(KEY_TYPES)
+    .filter(([key]) => key.startsWith(type))
+    .map(([key, algorithm]) => {
+      const [, prefix, length] = key.split('_')
+      const paddedAlgo = algorithm.padEnd(colWidth.algorithm)
+      const paddedPrefix = prefixRepr(Number(prefix)).padEnd(colWidth.prefix)
+      return `${paddedAlgo} - Prefix: ${paddedPrefix} Length: ${length} bytes`
+    })
+    .join('\n')
+}
+
+function keyError({
+  key,
+  type,
+  prefix,
+  len,
+}: {
+  key: string
+  type: KeyType
+  prefix: number
+  len: number
+}) {
+  const validFormats = getValidFormatsTable(type)
+
+  return `invalid_key:
+
+Type: ${type}
+Key: ${key}
+Prefix: ${prefixRepr(prefix)} 
+Length: ${len} bytes
+
+Acceptable ${type} formats are:
+${validFormats}
+`
+}
+
+/**
+ * Determines the algorithm associated with a given key (public/private).
+ *
+ * @param key - hexadecimal string representation of the key.
+ * @param type - whether expected key is public or private
+ * @returns Algorithm algorithm for signing/verifying
+ * @throws Error when key is invalid
+ */
+export function getAlgorithmFromKey(key: HexString, type: KeyType): Algorithm {
+  const { prefix, len } = getKeyInfo(key)
+  // Special case back compat support for no prefix
+  const usedPrefix = type === 'private' && len === 32 ? Prefix.NONE : prefix
```

Comment:
Are there syntax issues with using `type` as a param name? I know that would cause problems in Python, not sure if JS is as picky.
Also, nit: it would be a bit neater if `private` and `public` were constants (`PRIVATE` and `PUBLIC`)

Reply:
`type` is not a reserved word but will take the other suggestion

Reply:
```typescript
function generateSeed(
  options: {
    entropy?: Uint8Array
    algorithm?: Algorithm
  } = {},
): string {
  assert.ok(
    !options.entropy || options.entropy.length >= 16,
    'entropy too short',
  )
  const entropy = options.entropy
    ? options.entropy.slice(0, 16)
    : randomBytes(16)
  const type = options.algorithm === 'ed25519' ? 'ed25519' : 'secp256k1'
  return addressCodec.encodeSeed(entropy, type)
}
```

In TypeScript, the use of literal types, as shown in the existing generateSeed function, significantly reduces the chance of making mistakes such as typos. Literal types provide a clear and type-safe way to specify exact value(s) a variable can have, which is checked at compile-time, ensuring correctness.

In some cases, it might make sense to define constants or enums, especially if a set of related values is used in many places throughout the codebase or if there's a good chance those values will be refactored in the future. However, in this particular scenario, there's no notable benefit to adding the ceremony of defining additional constants or types. The direct use of string literals here is straightforward, adheres to existing code conventions, and keeps the code simple and easy to understand without losing the benefits of type safety.